amethyst_network/src/simulation/transport.rs

```diff
@@ -0,0 +1,176 @@
+//! This module holds the underlying system implementations for each of the various transport
+//! protocols. One important thing to note if you're implementing your own, the underlying sockets
+//! MUST be non-blocking in order to play nicely with the ECS scheduler.
+
+pub mod laminar;
+pub mod socket;
+pub mod udp;
+
+const NETWORK_SIM_TIME_SYSTEM_NAME: &str = "simulation_time";
+const NETWORK_SEND_SYSTEM_NAME: &str = "network_send";
+const NETWORK_RECV_SYSTEM_NAME: &str = "network_recv";
+const NETWORK_POLL_SYSTEM_NAME: &str = "network_poll";
+
+use crate::simulation::{
+    transport::socket::Socket, Message, NetworkSimulationResource, NetworkSimulationTime,
+    UrgencyRequirement,
+};
+use log::warn;
+use std::net::SocketAddr;
+
+/// Shared set up code for implementations of `NetworkSendSystem`s
+pub fn run_network_send_system<T: Socket>(
+    net: &mut NetworkSimulationResource<T>,
+    sim_time: &NetworkSimulationTime,
+    mut handle_send: impl FnMut(&mut T, SocketAddr, &Message) -> (),
+) {
+    // If no socket configured, this system should be a no-op.
+    if !net.has_socket() {
+        if net.has_messages() {
+            warn!("Messages waiting to be sent but no socket configured.");
+        }
+        return;
+    }
+
+    let messages = net.drain_messages(|message| {
+        message.urgency == UrgencyRequirement::Immediate
+            || sim_time.should_send_message_this_frame()
+    });
+
+    // If we have no messages to send, we're done here.
+    if messages.is_empty() {
+        return;
+    }
+
+    // If we're a server, we need to broadcast messages to all connected clients. If we're
+    // just a client, we're just going to send to our designated server.
+    if net.is_server() {
+        let clients = net.clients().to_vec();
+        let socket = net.get_socket_mut().expect("A socket should be configured");
+        clients.iter().map(|client| client.addr()).for_each(|addr| {
+            for message in messages.iter() {
+                handle_send(socket, addr, message);
+            }
+        });
+    } else {
+        let server_addr = net.server_addr();
+        let socket = net.get_socket_mut().expect("A socket should be configured");
```

Comment:
I actually couldn't figure out how to do this because it requires that I have a mutable reference to the net. Once I get one of those, I can't call `net.server_addr()` or `net.clients()`